### PR TITLE
Reduce Technical Debt. Hide details.

### PR DESCRIPTION
**Pull request by 3Hren**
WIP.

### CODE REVIEW

File: node/CMakeLists.txt [hunk 1]
```diff
@@ -27,7 +27,7 @@ add_library(node SHARED
     src/module.cpp
     src/node.cpp
     src/node/app.cpp
-    src/node/app/stats.cpp
+    src/node/stats
```
Comment (antmat) [inline node/CMakeLists.txt]: order, extension

File: node/CMakeLists.txt [hunk 2]
```diff
@@ -39,13 +39,17 @@ add_library(node SHARED
...
     src/node/slave/error.cpp
     src/node/slave/fetcher.cpp
-    src/node/slave/state/active.cpp
-    src/node/slave/state/handshaking.cpp
-    src/node/slave/state/sealing.cpp
-    src/node/slave/state/spawning.cpp
-    src/node/slave/state/state.cpp
-    src/node/slave/state/stopped.cpp
-    src/node/slave/state/terminating.cpp
+    src/node/engine
```
Comment (antmat) [inline node/CMakeLists.txt]: order

File: node/CMakeLists.txt [hunk 3]
```diff
@@ -39,13 +39,17 @@ add_library(node SHARED
...
+    src/node/slave/id
+    src/node/slave/machine
+    src/node/slave/state/active
+    src/node/slave/state/handshake
+    src/node/slave/state/inactive
+    src/node/slave/state/seal
+    src/node/slave/state/spawn
+    src/node/slave/state/state
+    src/node/slave/state/terminate
+    src/node/slave/stats
```
Comment (antmat) [inline node/CMakeLists.txt]: extensions :-)

File: node/include/cocaine/api/stream.hpp [hunk 4]
```diff
@@ -1,33 +1,26 @@
...
 #include <system_error>
 
 namespace cocaine {
 namespace hpack {
+
 class header_storage_t;
-}} // namespace cocaine::hpack
+
+}  // namespace hpack
+}  // namespace cocaine
```
Comment (antmat) [inline node/include/cocaine/api/stream.hpp:12]: Do we have an agreement on style of enclosed namespaces?

**Comment by 3Hren** [comment]
- [x] Extensions in CMakeLists.

File: node/include/cocaine/api/stream.hpp (hunk 4 above)
Comment (antmat) [inline node/include/cocaine/api/stream.hpp:12]: Or clang-format handle that too?

File: node/include/cocaine/detail/service/node/dispatch/handshake.hpp [hunk 5]
```diff
@@ -9,11 +9,12 @@
 
 #include "cocaine/idl/rpc.hpp"
 
-#include "cocaine/service/node/slot.hpp"
+#include "cocaine/detail/service/node/forwards.hpp"
```
Comment (antmat) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:12]: order of includes

Comment (3Hren) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:12]: What's wrong here?

Comment (antmat) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:12]: "detail" goes before "idl"?

Comment (3Hren) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:12]: I have the following logic of includes order: from common to details.
1. Standard POSIX or C.
2. Standard C++.
3. Boost.
4. External libraries.
5. Our external libraries.
6. Local libraries.
7. Current library public.
8. Current library details.

Comment (antmat) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:12]: That makes sense, however it is not consistent with cocaine style. We need to make an agreement, that in all new code we will use that new convention.

Comment (3Hren) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:12]: This allows for us locally to rewrite or undef macro for example.

Comment (3Hren) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:12]: Should we organize a committee?

Comment (antmat) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:12]: You can start a CodeStyle.md in root of plugins and write it there.

File: node/include/cocaine/detail/service/node/dispatch/handshake.hpp [hunk 6]
```diff
@@ -9,11 +9,12 @@
...
+#include "cocaine/detail/service/node/rpc/slot.hpp"
 
 namespace cocaine {
 
-class control_t;
+using detail::service::node::slave::control_t;
```
Comment (antmat) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:17]: Maybe local class typedef will be better?

**Comment by 3Hren** [comment]
- [x] Code Style File.

File: node/include/cocaine/detail/service/node/engine.hpp [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    /// The event will be put into the queue if there are no slaves available at this moment or all
+    /// of them are busy.
+    ///
+    /// \return the dispatch object, which is ready for processing the appropriate protocol
+    /// messages.
+    ///
+    /// \param downstream represents the [Client <- Worker] stream.
+    /// \param event an invocation event.
+    /// \param id represents slave id to be enqueued (may be none, which means any slave).
+    std::shared_ptr<client_rpc_dispatch_t> enqueue(
```
Comment (antmat) [inline node/include/cocaine/detail/service/node/engine.hpp]: I don't get the purpose of "auto smth() -> void", but better to use one style.
Or describe when we use one or another.

File: node/include/cocaine/detail/service/node/slave/machine.hpp [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+namespace cocaine {
+namespace detail {
+namespace service {
+namespace node {
+namespace slave {
+
+using cocaine::service::node::slave::id_t;
+
+using detail::service::node::slave::channel_t;
```
Comment (antmat) [inline node/include/cocaine/detail/service/node/slave/machine.hpp]: It's the same namespace, isn't it?

File: node/include/cocaine/detail/service/node/slave/machine.hpp [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+using detail::service::node::slave::control_t;
+using detail::service::node::slave::fetcher_t;
+using detail::service::node::slave::load_t;
+using detail::service::node::slave::state::state_t;
+
+/// Actual slave implementation.
+class machine_t : public std::enable_shared_from_this<machine_t> {
+    friend class detail::service::node::slave::state::active_t;
```
Comment (antmat) [inline node/include/cocaine/detail/service/node/slave/machine.hpp]: here too.

File: node/include/cocaine/detail/service/node/engine.hpp (hunk 7 above)
Comment (3Hren) [inline node/include/cocaine/detail/service/node/engine.hpp]: Ok, let's do `auto do_(Args...) -> R`. Looks ultra cyber cool, modern C++, yaw!

Comment (antmat) [inline node/include/cocaine/detail/service/node/engine.hpp]: Only application I can find is using decltype of parameter as return type, but ok, let's use this cyber cool, modern C++!

File: node/include/cocaine/detail/service/node/slave/machine.hpp [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    friend class detail::service::node::slave::state::handshake_t;
+    friend class detail::service::node::slave::state::seal_t;
+    friend class detail::service::node::slave::state::spawn_t;
+    friend class detail::service::node::slave::state::inactive_t;
+    friend class detail::service::node::slave::state::terminate_t;
+
+    friend class client_rpc_dispatch_t;
+
+    friend class detail::service::node::slave::channel_t;
```
Comment (antmat) [inline node/include/cocaine/detail/service/node/slave/machine.hpp]: here you already have "using"

File: node/include/cocaine/detail/service/node/slave/machine.hpp [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+    /// The flag means that the overseer has been destroyed and we shouldn't call the callback.
+    std::atomic<bool> closed;
+    cleanup_handler cleanup;
+
+    splitter_t splitter;
+    synchronized<std::shared_ptr<fetcher_t>> fetcher;
+    boost::circular_buffer<std::string> lines;
+
+    std::atomic<bool> shutdowned;
```
Comment (antmat) [inline node/include/cocaine/detail/service/node/slave/machine.hpp:79]: why not atomic_flag?

File: node/include/cocaine/detail/service/node/engine.hpp (hunk 7 above)
Comment (3Hren) [inline node/include/cocaine/detail/service/node/engine.hpp]: The same shit. Actually we can't use "real" modern C++, so this part of the language sweets my heart.

File: node/include/cocaine/detail/service/node/dispatch/handshake.hpp (hunk 6 above)
Comment (3Hren) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:17]: What's the difference?

Comment (antmat) [inline node/include/cocaine/detail/service/node/dispatch/handshake.hpp:17]: "using" is exposing it to whole cocaine namespace.

File: node/include/cocaine/detail/service/node/slave/machine.hpp (hunk 11 above)
Comment (3Hren) [inline node/include/cocaine/detail/service/node/slave/machine.hpp:79]: ¯_(ツ)_/¯

File: node/include/cocaine/detail/service/node/slave/state/handshake.hpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
+#pragma once
```
Comment (antmat) [inline node/include/cocaine/detail/service/node/slave/state/handshake.hpp]: I don't like this name change.
First of all I'll have the whole file in conflict. 
Second - state "handshaking" sound ok, state "handshake" is not any better.

Comment (3Hren) [inline node/include/cocaine/detail/service/node/slave/state/handshake.hpp]: Alternatives?

Comment (antmat) [inline node/include/cocaine/detail/service/node/slave/state/handshake.hpp]: leave it handshaking_t and handshaking.hpp ?

Comment (antmat) [inline node/include/cocaine/detail/service/node/slave/state/handshake.hpp]: besides it's really hard to compare diffs)

Comment (3Hren) [inline node/include/cocaine/detail/service/node/slave/state/handshake.hpp]: Ok, if you ask. But you don't ask with respect.
image

File: node/include/cocaine/detail/service/node/slave/state/seal.hpp [hunk 13]
```diff
@@ -4,21 +4,20 @@
```
Comment (antmat) [inline node/include/cocaine/detail/service/node/slave/state/seal.hpp:4]: Do you know why git handles renaming correctly here, but not in other cases?

Comment (3Hren) [inline node/include/cocaine/detail/service/node/slave/state/seal.hpp:4]: ¯_(ツ)_/¯

File: node/include/cocaine/detail/service/node/slave/machine.hpp (hunk 11 above)
Comment (3Hren) [inline node/include/cocaine/detail/service/node/slave/machine.hpp:79]: Personally I don't like atomics, but here is the right place not to touch ever no matter what.

File: node/include/cocaine/service/node.hpp [hunk 14]
```diff
@@ -18,86 +18,60 @@
...
+}  // namespace service
+}  // namespace cocaine
 
-auto
-node_category() -> const std::error_category&;
+namespace cocaine {
+namespace error {
 
-}} // namespace cocaine::error
+auto node_category() -> const std::error_category&;
```
Comment (antmat) [inline node/include/cocaine/service/node.hpp:74]: BTW - registerit in registrar in module initialization.

**Comment by antmat** [comment]
НБМХ

**Pull request merged**